# Lab: Pandas

Last week we loaded data using [`pandas`](https://pandas.pydata.org/)' `read.csv()` method, but Pandas can do way more than that. Pandas is an essential library for data science, as it provides the data structures (nameley, `series` -1D- and `data frames` -2D) and operations for manipulating tabular data. 

In this unit's labs we will be using `pandas` to read in, process and explore data (this notebook); create (basic) visualisations capabilities (@sec-pandas-datavis) as well as transforming data (@sec-pandas-transforming-data). To do so, we will be using a dataset about _"The Office"_.

## Dataset

[_The Office_](https://www.imdb.com/title/tt0386676/?ref_=ttep_ov) is a humoristic TV series originally created in 2001 by Ricky Gervais and Stephen Merchant that has received several adaptations. The dataset that we will be using contains information (i.e., title, date and ratings from [IMBD](https://www.imdb.com/title/tt0386676/?ref_=ttep_ov)) about every episode of the 9 seaons of the very successful USA's adaptation aired between 2005 and 2013.


::: aside
![_The Office_ promotional poster](img/the_office.png){width=400}
:::

The dataset is stored in a `csv` file that has the following columns: `season`, `episode`, `title`, `imdb_rating`, `total_votes`, `air_date`.


## Starting

To work with the dataset we will need to _import_ pandas so we can use every feature provided by the library, as well as loading the dataset stored in the `office_ratings.csv`.


In [1]:
# These two lines are added so that not all the warnings are rendered in the cell. 
# We do this not to confuse you during your learning journey with some of the warnings but 
# normally you would want them turned on since they can tell you something about things that might not be working as expected.
import warnings
warnings.filterwarnings('ignore')

In [2]:
# this following import will always be needed whenever you want to work with Pandas.
import pandas as pd

df = pd.read_csv('data/raw/office_ratings.csv', encoding='UTF-8')

In [3]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 188 entries, 0 to 187
Data columns (total 6 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   season       188 non-null    int64  
 1   episode      188 non-null    int64  
 2   title        188 non-null    str    
 3   imdb_rating  188 non-null    float64
 4   total_votes  188 non-null    int64  
 5   air_date     188 non-null    str    
dtypes: float64(1), int64(3), str(2)
memory usage: 8.9 KB


## Help!

Python has inbuilt documentation. To access this add a `?` before an object or method.

::: callout-note

The output of the help function has been omitted in the handbook. Please run the cells in your notebook to read the different outputs

:::

For example, our dataframe

In [4]:
?df

or the `dtypes` property

::: callout-tip

Properties of object are values associated with the object and are not called with a `()` at the end.

:::

In [5]:
?df.dtypes

The `info` method for dataframes.

In [6]:
?df.info

If you would like to get help in-line like the examples above, that can give a very long help message that might not be always convenient. If you like, you can try to get the help for this following function like this:

> ?pd.read_csv

However, the below will be quite long -- it provides you the various arguments (options) you can use with the method. 

Instead of this approach, a much better way to get help is to refer to the documentation and the API of the library that you are using. For instance, for `read_csv()`, this page is much more useful -- https://pandas.pydata.org/docs/reference/api/pandas.read_csv.html 

We recommend that you use a search engine very frequently.

In [7]:
?pd.read_csv

The Pandas documentation is rather good. Relevent to our below work is:

* [What kind of data does pandas handle?](https://pandas.pydata.org/pandas-docs/stable/getting_started/intro_tutorials/01_table_oriented.html#min-tut-01-tableoriented)
* [How to calculate summary statistics?](https://pandas.pydata.org/pandas-docs/stable/getting_started/intro_tutorials/01_table_oriented.html#min-tut-01-tableoriented)
* [How to create plots in pandas?](https://pandas.pydata.org/pandas-docs/stable/getting_started/intro_tutorials/04_plotting.html#min-tut-04-plotting)
* [How to handle time series data with ease?](https://pandas.pydata.org/pandas-docs/stable/getting_started/intro_tutorials/09_timeseries.html#min-tut-09-timeseries)

I also found [a rather nice series of lessons a kind person put together](https://bitbucket.org/hrojas/learn-pandas/src/master/). There are lots of online tutorials which will help you.

## Structure

In @sec-data-types we introduced Python's data types as well as how to use the function `type()` to retrieve an object's data type. Pandas expands python's data types by creating a new one called `data frame` 

::: aside

Do you remember what are Python's data types? You can refer to @sec-data-types for a refresher and to know more about them.

:::

::: callout-tip

### Data frames

Data frames are 2-dimensional data structures that store information in columns and rows, very much like data is stored in a spreadsheet or a database. Typically, every column will contain variables (or sometimes called attributes) whereas every row represents an observation. This is known as wide data frames, as opposed to long data frames.

In pandas, every column has a name and rows can be named, too. 

:::

So let's check the what our newly created object's (`df`) data type:


In [8]:
type(df)

pandas.DataFrame

Unsurprisingly, `df` is a `DataFrame`` object, provided by pandas.

The `DataFrame` object has lots of built in _methods_ and _attributes_.

The `info` method gives us information about datatypes, dimensions and the presence of null values in our dataframe. Let's see how can we use it and what information is returned:

In [9]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 188 entries, 0 to 187
Data columns (total 6 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   season       188 non-null    int64  
 1   episode      188 non-null    int64  
 2   title        188 non-null    str    
 3   imdb_rating  188 non-null    float64
 4   total_votes  188 non-null    int64  
 5   air_date     188 non-null    str    
dtypes: float64(1), int64(3), str(2)
memory usage: 8.9 KB


We can just `dtypes` to check the data types of every variable in the data frame.

In [10]:
df.dtypes

season           int64
episode          int64
title              str
imdb_rating    float64
total_votes      int64
air_date           str
dtype: object

Or just the dimensions (e.g., rows and columns).

In [11]:
df.shape

(188, 6)

In this case, there are only 188 rows. But for larger datasets we might want to look at the head (top 5) and tail (bottom 5) rows using `.head()` and `.tail()`, respectively.

In [12]:
df.head()

,season,episode,title,imdb_rating,total_votes,air_date
0,1,1,Pilot,7.6,3706,2005-03-24
1,1,2,Diversity Day,8.3,3566,2005-03-29
2,1,3,Health Care,7.9,2983,2005-04-05
3,1,4,The Alliance,8.1,2886,2005-04-12
4,1,5,Basketball,8.4,3179,2005-04-19


In [13]:
df.tail()

,season,episode,title,imdb_rating,total_votes,air_date
183,9,19,Stairmageddon,8.0,1484,2013-04-11
184,9,20,Paper Airplane,8.0,1482,2013-04-25
185,9,21,Livin' the Dream,8.9,2041,2013-05-02
186,9,22,A.A.R.M.,9.3,2860,2013-05-09
187,9,23,Finale,9.7,7934,2013-05-16


## Summary

To get an overview of our data we can ask Python to '_describe_ our (numeric) data'

In [14]:
df.describe()

,season,episode,imdb_rating,total_votes
count,188.000000,188.000000,188.000000,188.000000
mean,5.468085,11.877660,8.257447,2126.648936
std,2.386245,7.024855,0.538067,787.098275
min,1.000000,1.000000,6.700000,1393.000000
25%,3.000000,6.000000,7.900000,1631.500000
50%,6.000000,11.500000,8.200000,1952.500000
75%,7.250000,18.000000,8.600000,2379.000000
max,9.000000,26.000000,9.700000,7934.000000


or we can pull out specific statistics for numeric columns.

In [15]:
#| error: true

df.mean()

TypeError: Cannot perform reduction 'mean' with string dtype

Note the error triggered above due to pandas attempting to calculate the mean of the wrong type (i.e. non-numeric values). We can address that by only computing the mean of numeric values (see below):

In [16]:
df.mean(numeric_only=True)

season            5.468085
episode          11.877660
imdb_rating       8.257447
total_votes    2126.648936
dtype: float64

or the sum of every value within the same column:

In [17]:
df.sum()

season                                                      1028
episode                                                     2233
title          PilotDiversity DayHealth CareThe AllianceBaske...
imdb_rating                                               1552.4
total_votes                                               399810
air_date       2005-03-242005-03-292005-04-052005-04-122005-0...
dtype: object

Similarly to what happened with `mean()`, `sum()` is adding all values in every observation of every attribute, regardless of their type, but this time is not producing an error. Can you see what happens with strings? And with dates?

Again, we can force to use numeric values only:

In [18]:
df.sum(numeric_only=True)

season           1028.0
episode          2233.0
imdb_rating      1552.4
total_votes    399810.0
dtype: float64

## Subsetting

Often times we may have a large dataset and we only need to work with just a part of it (a subset) consisting of certain columns and/or rows. Selecting specific columns and/or rows is known as subsetting.

### Selecting columns

Because in pandas every column has a name, we can select columns by their name or their position.

#### Selecting by name

To select by name we will use the syntax `df['<column_name>']`. For example, if we wanted to select the ratings:

In [19]:
df['imdb_rating']

0      7.6
1      8.3
2      7.9
3      8.1
4      8.4
      ... 
183    8.0
184    8.0
185    8.9
186    9.3
187    9.7
Name: imdb_rating, Length: 188, dtype: float64

or we could select the date in which the chapters were first aired:

In [20]:
df['air_date']

0      2005-03-24
1      2005-03-29
2      2005-04-05
3      2005-04-12
4      2005-04-19
          ...    
183    2013-04-11
184    2013-04-25
185    2013-05-02
186    2013-05-09
187    2013-05-16
Name: air_date, Length: 188, dtype: str

We can even select more than one column!

In [21]:
df[['imdb_rating', 'total_votes']]

,imdb_rating,total_votes
0,7.6,3706
1,8.3,3566
2,7.9,2983
3,8.1,2886
4,8.4,3179
...,...,...
183,8.0,1484
184,8.0,1482
185,8.9,2041
186,9.3,2860


::: callout-important

Did you notice that we used two sets of squared brackets (`[[]]`)? This is needed because we need to passing a _list_ of the column names to the [`__getitem__`](https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#basics) method of the pandas dataframe object, and as you may remember from @sec-list-dictionaries, this is the syntax used for lists (thank [this stackoverflow question](https://stackoverflow.com/questions/11285613/selecting-multiple-columns-in-a-pandas-dataframe)). 

This is what we'd get otherwise:

In [22]:
#| error: true
df['imdb_rating', 'total_votes']

KeyError: ('imdb_rating', 'total_votes')

You can also check out the pandas documentation on [indexing and selecting data](https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#basics).
:::

We can also apply methods to subset, such as this one to get the average rating:

In [23]:
df['imdb_rating'].mean()


np.float64(8.25744680851064)

Or to calculate the total number of votes:

In [24]:
df['total_votes'].sum()

np.int64(399810)

Or a combination of multiple columns:

In [25]:
df[['imdb_rating', 'total_votes']].mean()

imdb_rating       8.257447
total_votes    2126.648936
dtype: float64

#### Selecting by position

If we do not want to use column names, we can use `iloc` method by using the syntax `<object>.iloc[<row slice>, <column slice>]`, where a slice is a range of numbers separated by a colon `:`. So, if we were to select the value in the 4th row and 2nd column, we'd use:

In [26]:
df.iloc[4,2]

'Basketball'

But if we just wanted to select a column? In that case, we can use the same method but instead of specifiying a row, we will need to use `:` to indicate that we are selecting all the rows, such as:

In [27]:
df.iloc[:,2]

0                 Pilot
1         Diversity Day
2           Health Care
3          The Alliance
4            Basketball
             ...       
183       Stairmageddon
184      Paper Airplane
185    Livin' the Dream
186            A.A.R.M.
187              Finale
Name: title, Length: 188, dtype: str

Conversely, if we just wanted to select all the columns from a given row, we'd use `:` on the right side of the `,` like this:

In [28]:
df.iloc[4,:]

season                  1
episode                 5
title          Basketball
imdb_rating           8.4
total_votes          3179
air_date       2005-04-19
Name: 4, dtype: object

We can use negative values in indexes to indicate 'from the end'. So, an index of [-10, :] returns the 10th from last row.

In [29]:
df.iloc[-10,:]

season                  9
episode                14
title           Vandalism
imdb_rating           7.6
total_votes          1402
air_date       2013-01-31
Name: 178, dtype: object

Instead of using tail, we could ask for the last 5 rows with an index of `[-5:, :]`. I read `:` as 'and everything else' in these cases.

In [30]:
df.iloc[-5:,:]

,season,episode,title,imdb_rating,total_votes,air_date
183,9,19,Stairmageddon,8.0,1484,2013-04-11
184,9,20,Paper Airplane,8.0,1482,2013-04-25
185,9,21,Livin' the Dream,8.9,2041,2013-05-02
186,9,22,A.A.R.M.,9.3,2860,2013-05-09
187,9,23,Finale,9.7,7934,2013-05-16


In [31]:
df.tail()

,season,episode,title,imdb_rating,total_votes,air_date
183,9,19,Stairmageddon,8.0,1484,2013-04-11
184,9,20,Paper Airplane,8.0,1482,2013-04-25
185,9,21,Livin' the Dream,8.9,2041,2013-05-02
186,9,22,A.A.R.M.,9.3,2860,2013-05-09
187,9,23,Finale,9.7,7934,2013-05-16


Note that the row is shown on the left. That will stop you getting lost in slices of the data. 

For the top ten rows

In [32]:
df.iloc[:10,:]

,season,episode,title,imdb_rating,total_votes,air_date
0,1,1,Pilot,7.6,3706,2005-03-24
1,1,2,Diversity Day,8.3,3566,2005-03-29
2,1,3,Health Care,7.9,2983,2005-04-05
3,1,4,The Alliance,8.1,2886,2005-04-12
4,1,5,Basketball,8.4,3179,2005-04-19
5,1,6,Hot Girl,7.8,2852,2005-04-26
6,2,1,The Dundies,8.7,3213,2005-09-20
7,2,2,Sexual Harassment,8.2,2736,2005-09-27
8,2,3,Office Olympics,8.4,2742,2005-10-04
9,2,4,The Fire,8.4,2713,2005-10-11


Of course, we can run methods on these slices. We could, if we wanted to, calculate the mean imdb rating of only the first and last 100 episodes. _Note_ the indexing starts at 0 so we want the column index of 3 (0:season, 1:episode, 2:title, 3:imdb_rating).

In [33]:
df.iloc[:100,3].mean()

np.float64(8.483)

In [34]:
df.iloc[-100:,3].mean()

np.float64(8.062)

If you are unsure how many rows you have then the count method comes to the rescue.

In [35]:
df.iloc[-100:,3].count()

np.int64(100)

In [36]:
df.describe()

,season,episode,imdb_rating,total_votes
count,188.000000,188.000000,188.000000,188.000000
mean,5.468085,11.877660,8.257447,2126.648936
std,2.386245,7.024855,0.538067,787.098275
min,1.000000,1.000000,6.700000,1393.000000
25%,3.000000,6.000000,7.900000,1631.500000
50%,6.000000,11.500000,8.200000,1952.500000
75%,7.250000,18.000000,8.600000,2379.000000
max,9.000000,26.000000,9.700000,7934.000000


So it looks like the last 100 episodes were less good than the first 100. I guess that is why it was cancelled.

Our data is organised by season. Looking at the average by season might help.

In [37]:
df[['season', 'imdb_rating']].groupby('season').mean()

,imdb_rating
season,
1,8.016667
2,8.436364
3,8.573913
4,8.600000
5,8.492308
6,8.219231
7,8.316667
8,7.666667
9,7.956522


The above line groups our dataframe by values in the season column and then displays the mean for each group. Pretty nifty.

Season 8 looks pretty bad. We can look at just the rows for season 8.

In [38]:
df[df['season'] == 8]

,season,episode,title,imdb_rating,total_votes,air_date
141,8,1,The List,8.2,1829,2011-09-22
142,8,2,The Incentive,8.2,1668,2011-09-29
143,8,3,Lotto,7.3,1601,2011-10-06
144,8,4,Garden Party,8.1,1717,2011-10-13
145,8,5,Spooked,7.6,1543,2011-10-27
146,8,6,Doomsday,7.8,1476,2011-11-03
147,8,7,Pam's Replacement,7.7,1563,2011-11-10
148,8,8,Gettysburg,7.0,1584,2011-11-17
149,8,9,Mrs. California,7.7,1553,2011-12-01
150,8,10,Christmas Wishes,8.0,1547,2011-12-08


### Filtering rows

We can filter rows matching some criteria by using the syntax `<object>.loc[<criteria>]`.  So, if we wanted to filter all the episodes from the 8th season, we would do the following:

In [39]:
df.loc[df['season'] == 8]

,season,episode,title,imdb_rating,total_votes,air_date
141,8,1,The List,8.2,1829,2011-09-22
142,8,2,The Incentive,8.2,1668,2011-09-29
143,8,3,Lotto,7.3,1601,2011-10-06
144,8,4,Garden Party,8.1,1717,2011-10-13
145,8,5,Spooked,7.6,1543,2011-10-27
146,8,6,Doomsday,7.8,1476,2011-11-03
147,8,7,Pam's Replacement,7.7,1563,2011-11-10
148,8,8,Gettysburg,7.0,1584,2011-11-17
149,8,9,Mrs. California,7.7,1553,2011-12-01
150,8,10,Christmas Wishes,8.0,1547,2011-12-08


::: callout-note

### Understanding the criteria

To understand why we have to write the name of the dataframe twice, we can focus on the output provided by the filtering criteria only:


In [40]:
df['season'] == 9

0      False
1      False
2      False
3      False
4      False
       ...  
183     True
184     True
185     True
186     True
187     True
Name: season, Length: 188, dtype: bool

As you can see, it returns a boolean series specifiying which rows are matching the criteria (`True`) and which ones are not (`False`)

As a side note, while writing the name of the dataframe twice may seem redundant, this means that we could filter rows based on other objects.
:::

We can get an overview of the rating of all chapters within season 8 by:

In [41]:
df.loc[df['season'] == 8, 'imdb_rating'].describe()

count    24.000000
mean      7.666667
std       0.405041
min       6.700000
25%       7.450000
50%       7.800000
75%       7.925000
max       8.200000
Name: imdb_rating, dtype: float64

Generally pretty bad, but there is clearly one very disliked episode.

## Adding columns

We can add new columns pretty simply.

In [42]:
df['x'] = 44
df.head()

,season,episode,title,imdb_rating,total_votes,air_date,x
0,1,1,Pilot,7.6,3706,2005-03-24,44
1,1,2,Diversity Day,8.3,3566,2005-03-29,44
2,1,3,Health Care,7.9,2983,2005-04-05,44
3,1,4,The Alliance,8.1,2886,2005-04-12,44
4,1,5,Basketball,8.4,3179,2005-04-19,44


Our new column can be an operation on other columns

In [43]:
df['rating_div_total_votes'] = df['imdb_rating'] / df['total_votes']
df.head()

,season,episode,title,imdb_rating,total_votes,air_date,x,rating_div_total_votes
0,1,1,Pilot,7.6,3706,2005-03-24,44,0.002051
1,1,2,Diversity Day,8.3,3566,2005-03-29,44,0.002328
2,1,3,Health Care,7.9,2983,2005-04-05,44,0.002648
3,1,4,The Alliance,8.1,2886,2005-04-12,44,0.002807
4,1,5,Basketball,8.4,3179,2005-04-19,44,0.002642


or as simple as adding one to every value.

In [44]:
df['y'] = df['season'] + 1
df.iloc[0:5,:]

,season,episode,title,imdb_rating,total_votes,air_date,x,rating_div_total_votes,y
0,1,1,Pilot,7.6,3706,2005-03-24,44,0.002051,2
1,1,2,Diversity Day,8.3,3566,2005-03-29,44,0.002328,2
2,1,3,Health Care,7.9,2983,2005-04-05,44,0.002648,2
3,1,4,The Alliance,8.1,2886,2005-04-12,44,0.002807,2
4,1,5,Basketball,8.4,3179,2005-04-19,44,0.002642,2


In [45]:
y =  df['season'] + 1

## Writing data

Pandas supports writing out data frames to various formats.

In [46]:
?df.to_csv

Now you can uncomment the code below to save your dataframe into a csv file. But before doing so, check that your `data/output` folder is empty, as it would override its content:

In [47]:
#| eval: false
#df.to_csv('data/output/my_output_ratings.csv', encoding='UTF-8')

Likewise, we could export our dataset to an excel file by using `to_excel`:

In [48]:
?df.to_excel

Now you can uncomment the code below to save your dataframe into an excel file. But before doing so, check that your `data/output` folder is empty:

In [49]:
# df.to_excel('data/output/my_output_ratings.xlsx')


## Combining datasets


In this notebook, our dataset was created from a single file that contained all the data that was needed. However, often times data will be spread into different files that we will need to combine to create our own dataset.

Consider the two dataframes below:

In [50]:
df_1 = pd.read_csv('data/raw/office1.csv', encoding='UTF-8')
df_2 = pd.read_csv('data/raw/office2.csv', encoding='UTF-8')

In [51]:
df_1.head()

,id,season,episode,imdb_rating
0,5-1,5,1,8.8
1,9-13,9,13,7.7
2,5-6,5,6,8.5
3,3-23,3,23,9.3
4,9-16,9,16,8.2


In [52]:
df_2.head()

,id,total_votes
0,4-10,2095
1,3-21,2403
2,7-24,2040
3,6-18,1769
4,8-8,1584


As can be seen, the total votes and imdb ratings data are split between files that we will need to combine. Usually this is done by using a shared column between the two datasets that works as an index. Gladly, `head()` reveals that in both cases there is a common column called `id`. We can _join_ the two dataframes together using the common column.

In [53]:
inner_join_office_df = pd.merge(df_1, df_2, on='id', how='inner')
inner_join_office_df

,id,season,episode,imdb_rating,total_votes
0,5-1,5,1,8.8,2501
1,9-13,9,13,7.7,1394
2,5-6,5,6,8.5,2018
3,3-23,3,23,9.3,3010
4,9-16,9,16,8.2,1572
...,...,...,...,...,...
183,5-21,5,21,8.7,2032
184,2-13,2,13,8.3,2363
185,9-6,9,6,7.8,1455
186,2-2,2,2,8.2,2736


In this way you can combine datasets using common columns and an _inner join_. We will leave that for the moment. If you want more information about merging data then see [this page](https://www.datasciencemadesimple.com/join-merge-data-frames-pandas-python/#:~:text=Merge%20%28%29%20Function%20in%20pandas%20is%20similar%20to,rows%20from%20both%20data%20frames%2C%20specify%20how%3D%20%E2%80%98outer%E2%80%99.) and the [pandas documentation](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.merge.html).

### Well done!

Well done! You've reached the end of a pretty long notebook that went through a lot of details about how to work with this pretty critical package called Pandas. We haven't really done a lot of detailed analysis in this one yet but we will use Pandas a lot and frequently.

Your best friend will be the Pandas documentation -- https://pandas.pydata.org/docs/index.html

This documentation is great. We particuarly recomment the User Guide that will answer most of your questions and will give you a lot of code to copy and paste first and then modify to do what you need to do -- https://pandas.pydata.org/docs/user_guide/index.html#user-guide
